# Data Analyst Professional Practical Exam Submission

**You can use any tool that you want to do your analysis and create visualizations. Use this template to write up your summary for submission.**

You can use any markdown formatting you wish. If you are not familiar with Markdown, read the [Markdown Guide](https://s3.amazonaws.com/talent-assets.datacamp.com/Markdown+Guide.pdf) before you start.


## 📝 Task List

Your written report should include written text summaries and graphics of the following:
- Data validation:   
  - Describe validation and cleaning steps for every column in the data 
- Exploratory Analysis:  
  - Include two different graphics showing single variables only to demonstrate the characteristics of data  
  - Include at least one graphic showing two or more variables to represent the relationship between features
  - Describe your findings
- Definition of a metric for the business to monitor  
  - How should the business use the metric to monitor the business problem
  - Can you estimate initial value(s) for the metric based on the current data
- Final summary including recommendations that the business should undertake

*Start writing report here..*

## ✅ When you have finished...
-  Publish your Workspace using the option on the left
-  Check the published version of your report:
	-  Can you see everything you want us to grade?
    -  Are all the graphics visible?
-  Review the grading rubric. Have you included everything that will be graded?
-  Head back to the [Certification Dashboard](https://app.datacamp.com/certification) to submit your practical exam report and record your presentation

## Data Understanding & Validation

In [1]:
#importing the necessary libraries
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

%matplotlib inline 

In [2]:
#reading in csv file
product_sales = pd.read_csv('product_sales.csv')

#displaying first few rows and info
display(product_sales.head())
print('\n')
display(product_sales.info())

week  sales_method                           customer_id  nb_sold  revenue  \
0     2         Email  2e72d641-95ac-497b-bbf8-4861764a7097       10      NaN   
1     6  Email + Call  3998a98d-70f5-44f7-942e-789bb8ad2fe7       15   225.47   
2     5          Call  d1de9884-8059-4065-b10f-86eef57e4a44       11    52.55   
3     4         Email  78aa75a4-ffeb-4817-b1d0-2f030783c5d7       11      NaN   
4     3         Email  10e6d446-10a5-42e5-8210-1b5438f70922        9    90.49   

   years_as_customer  nb_site_visits      state  
0                  0              24    Arizona  
1                  1              28     Kansas  
2                  6              26  Wisconsin  
3                  3              25    Indiana  
4                  0              28   Illinois



<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15000 entries, 0 to 14999
Data columns (total 8 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   week               15000 non-null  int64  
 1   sales_method       15000 non-null  object 
 2   customer_id        15000 non-null  object 
 3   nb_sold            15000 non-null  int64  
 4   revenue            13926 non-null  float64
 5   years_as_customer  15000 non-null  int64  
 6   nb_site_visits     15000 non-null  int64  
 7   state              15000 non-null  object 
dtypes: float64(1), int64(4), object(3)
memory usage: 937.6+ KB


None

In this dataset have a total of 15,000 entries and 8 columns. Only one column has null values - `revenue`.

In [3]:
#amount of null values for the revenue column  
product_sales['revenue'].isna().sum()

1074

The column, `revenue`, is missing 1,074 from a total of 15,000. The missing values accounts for 7.16% of the data. Will impute the missing values instead of deleting them because the dataset is already small. Most likely will be the median value due to the mean being susceptible to outliers.

### Examination of Each Attribute

In [4]:
#examining the week attribute 
product_sales['week'].value_counts()

1    3721
4    2575
5    2574
2    2491
3    2411
6    1228
Name: week, dtype: int64

`week`: week the sale was made, counted as week since product launch. There are 6 unique values. From here can see that the most sales are made in the 1st week of the product launch, with the 6th week having the lowest sales. However, the amount of sales does not follow the time chronologically with the 4th week since the product launch being the second highest and the 3rd week being the second lowest.

In [5]:
#examining sales_method attribute
product_sales['sales_method'].value_counts()

Email           7456
Call            4962
Email + Call    2549
em + call         23
email             10
Name: sales_method, dtype: int64

`sales_method`: sales method used for the customer. There are 3 methods. However there has to be some cleaning up since em + call can go under Email + Call and email goes under Email. Though even with changes can see that Email will still be the highest sale method used on a customer with the Call method in second at 4,962 - a 2,504 difference.

In [6]:
#correcting the values in sales_method
product_sales['sales_method'] = product_sales['sales_method'].replace('em + call', 'Email + Call')
product_sales['sales_method'] = product_sales['sales_method'].replace('email', 'Email')

#sanity check
product_sales['sales_method'].value_counts()

Email           7466
Call            4962
Email + Call    2572
Name: sales_method, dtype: int64

In [8]:
display(product_sales['customer_id'].head())

display(product_sales['customer_id'].tail())

0    2e72d641-95ac-497b-bbf8-4861764a7097
1    3998a98d-70f5-44f7-942e-789bb8ad2fe7
2    d1de9884-8059-4065-b10f-86eef57e4a44
3    78aa75a4-ffeb-4817-b1d0-2f030783c5d7
4    10e6d446-10a5-42e5-8210-1b5438f70922
Name: customer_id, dtype: object

14995    17267b41-d048-4346-8b90-7f787690a836
14996    09e10d6f-4508-4b27-895e-4db11ce8302b
14997    839653cb-68c9-48cb-a097-0a5a3b2b298b
14998    e4dad70a-b23b-407c-8bd3-e32ea00fae17
14999    4e077235-7c17-4054-9997-7a890336a214
Name: customer_id, dtype: object

Since `customer_id` is a unique identifier for the customer and there are no missing values, no cleaning is needed.

In [9]:
#examing the nb_sold attribute
product_sales['nb_sold'].value_counts()

10    3677
9     3535
11    2439
8     1941
12    1088
13     939
7      698
15     555
14      97
16      31
Name: nb_sold, dtype: int64

`nb_sold`: number of new products sold. There are 10 unique values. With a max of 16 new products sold for a customer. To maximize the number of new products sold can futher explore those values of `nb_sold`. However when looking at the highest amount of products sold - 10 there is a big difference between the amount of customers which bought 10 versus 16. 9 is a close second.

In [10]:
#examing the years_as_customer attribute
display(product_sales['years_as_customer'].value_counts())
display(len(product_sales['years_as_customer'].value_counts()))

1     2504
2     1987
3     1619
0     1471
4     1338
5     1118
6      918
7      709
8      608
9      509
10     405
11     318
12     284
13     239
14     167
15     150
16     123
17      88
18      82
19      58
20      54
22      43
21      37
24      27
26      22
23      21
25      19
27      15
30       9
34       8
33       8
28       8
31       6
35       5
29       5
32       5
36       4
39       3
37       2
38       2
47       1
63       1
Name: years_as_customer, dtype: int64

42

`years_as_customer`: number of years customer has been buying from Pens and Printers, the company was founded in 1984. At the time of this analysis it is 2023, 1984 was 39 years ago. Therefore there are some values that do not fit within the correct range. More customers tend to be newer customers with the higher amount of customers being within the 0 to 4 years as customers with 1 being the highest at 2,504 customers.  


In [11]:
#instance where the years_as_customer value is greater than 39
product_sales[product_sales['years_as_customer'] > 39]

week sales_method                           customer_id  nb_sold  \
13741     2        Email  18919515-a618-430c-9a05-2c7d8fea96af       10   
13800     4         Call  2ea97d34-571d-4e1b-95be-fea1c404649f       10   

       revenue  years_as_customer  nb_site_visits       state  
13741    97.22                 63              24  California  
13800    50.47                 47              27  California

There are only 2 instances that are greater than 39, will remove these since there is no right way of imputing these values unless I use the mode but again there are only 2. Therefore I will just remove these values.

In [12]:
#dropping instances with year greater than 39
product_sales.drop(product_sales[product_sales['years_as_customer'] > 39].index, inplace=True)

#sanity check
product_sales[product_sales['years_as_customer'] > 39]

Empty DataFrame
Columns: [week, sales_method, customer_id, nb_sold, revenue, years_as_customer, nb_site_visits, state]
Index: []

In [ ]:
site_visits = product_sales['nb_site_visits'].value_counts()

display(site_visits)
display(len(site_visits))
display(max(site_visits))
display(min(site_visits))


In [ ]:
product_sales[product_sales['years_as_customer'] > 39]

Since only two instances are greater than 39 and I don't really have a way of finding out the correct year of the customer will just remove these 2. 

In [ ]:
#sanity check
product_sales['years_as_customer'].value_counts()

In [ ]:
#examing the nb_site_visits attribute
product_sales['nb_site_visits'].value_counts()

In [ ]:
#examining the customer_id attribute to make sure all the customer_id are unique
len(product_sales['customer_id'].unique())

This is the right amount since dropped those 2 rows whose year did not match. 

In [ ]:
#examing the state attribute
display(product_sales['state'].value_counts())

display(len(product_sales['state'].unique()))

Look at distributions
Look at relationship between sales_method and revenue
impute missing values for revenue 
look at relationship between years_as_customer and revenue
groupby on sales_method and revenue

In [ ]:
#description of dataframe 
product_sales.describe()

In [ ]:
#missing values for revenue before imputation
sns.histplot(data=product_sales, x='revenue');

Can see that the bulk of values are on the lower end - 50 to a 100.

In [ ]:
#grabbing the rows which have the missing values of revenue in setting it to another dataframe to see which methods have the highest value count
missing = product_sales[product_sales['revenue'].isna()]

missing

In [ ]:
#checking value count of sales_method in missing dataframe
#unnormalized
display(missing['sales_method'].value_counts())

#normalized
display(missing['sales_method'].value_counts(normalize=True))

From here can see that the missing values of revenue are more frequent with the sales_method Email at 50.7%. However, email is the most frequently used sales_method according to the missing dataframe and the product_sales dataframe which could be why it has the most missing revenue values.

In [ ]:
product_sales['revenue'].value_counts()

In [ ]:
#imputing the missing values of revenue with the median value
product_sales['revenue'] = product_sales['revenue'].fillna(product_sales['revenue'].median())

#sanity check
product_sales.isna().sum()

In [ ]:
#checking specific customer_id
product_sales[product_sales['customer_id'] == '2e72d641-95ac-497b-bbf8-4861764a7097']

Null values have been replaced with the median of revenue which was 89.5. The reason the median was chosen is because don't want the data to be skewed towards potential outliers of the revenue.

In [ ]:
#grouping revenue by sales method
product_sales.groupby('sales_method')['revenue'].agg(['sum', 'median', 'mean'])

The sum of the revenue is the highest for Email which lines up since it the most frequently used sales method. However, for the the others - median and mean - Email + Call have the highest revenue which is an interesting find. Putting in the effort to use call and email seems to pay off. 

In [ ]:
product_sales.groupby('week')['revenue'].agg(['sum', 'median', 'mean'])

Here we can see that the earlier after the product launch the more sales happen. It is interesting to see that that they lower at week 3 they rise again at week 4. With week 5 being the second highest after week 1 in total revenue. The total revenue drops again at week 6. 

Another interesting takeway here is the median is highest for week 6 as well as the mean (which could be do an outlier). The median tends to rise with the week; except for at week 3 where there is a lull in revenue. The same is said for the mean of the revenue. 

TO DO List for 02/23
- look at the relationship between revenue and the each of the other columns (do the top 5 for the customers and the last ones) 
- relationship between customers and state
- `nb_sold`: numeric, number of new products sold
- `week`: week sale was made, counted as week since product launch
- `nb_site_visits`: numeric, number of times the customer has visited the website in the past 6 months 
- also fix the years since there is some that does not make sense

In [ ]:
#plotting the relationship with the top 5 funder groups and status group  
plt.figure(figsize=(15,7))
sns.barplot(data=product_sales, x='week', y='revenue', hue='sales_method')
plt.title('Revenue By Week Sales Method');

As shown in the graph above revenue, increases as the weeks progress and the revenue is higher with the sale method of Email + Call. 

In [ ]:
plt.figure(figsize=(15,7))
sns.scatterplot(data=product_sales, x='years_as_customer', y='revenue', hue='sales_method')
plt.title('Revenue By Years as Customer -  Sales Method');

Although this is not a very good visual. This is quite interesting to see the revenue  difference for the sales method due to the obivious separation between them. With call sales method bringing in the less revenue. 

In [ ]:
plt.figure(figsize=(15,7))
sns.barplot(data=product_sales, x='years_as_customer', y='revenue', hue='sales_method', order=product_sales.years_as_customer.value_counts().iloc[:10].index)
plt.title('Revenue By Years as Customer -  Sales Method');

From this chart can see that newer customers tend to spend more with 1, 2, and 3 not having much of a difference between them. Let's look at the bottom ten years.

In [ ]:
plt.figure(figsize=(15,7))
sns.barplot(data=product_sales, x='years_as_customer', y='revenue', hue='sales_method', order=product_sales.years_as_customer.value_counts().iloc[-10:].index)
plt.title('Revenue By Years as Customer -  Sales Method');

Here we can see that older customers have the lesser revenues. Another interesting find is that most of these execpt for years - 29, 39, &, 37 - the sales method Email + Call are not used. Which I found out previously that the method brings in more revenue than the others. 

In [ ]:
plt.figure(figsize=(15,7))
sns.barplot(data=product_sales, x='state', y='revenue', hue='sales_method', order=product_sales.state.value_counts().iloc[:10].index)
plt.title('Revenue By State -  Sales Method');

Shown from the chart can see that California has the most sales. Also can see that Email+Call tends to bring in the highest revenue once again; therefore this is a good sales method to focus on when bringing this into the recommendations.

In [ ]:
product_sales.groupby('state')['revenue'].agg(['sum', 'median', 'mean'])

Can see that in the sum the state with largest revenue is California, for the median the state is South Dakota, and for the mean is is also South Dakota. South Dakota has a low sum; however, with 4249.19. 

In [ ]:
product_sales.groupby('state')['years_as_customer'].agg('sum').sort_values(ascending=False)

Looking at this groupby can see that the top 5 states are California, Texas, New York, Florida, and Pennsylvania for how the amount of time their customers have been their customers. This list almost matches exactly with what the revenue is by state - cutt by sales method.

In [ ]:
state = product_sales.groupby('state')

cali = state.get_group('California')

cali['years_as_customer'].value_counts()

Most of the customers from California - the state that has most of the customers and highest revenue - have been their customer in 0 and 6 years. With most being a customer for one year with 323 a big difference between the two years as customers at 250. However, what can be done to keep older customers as interested in willing to pay more? Maybe the sales method need to change?

In [ ]:
#customer_id, years_as_customer, state
product_sales.head()

In [ ]:
#relationship with week, years_as_customer, revenue, and state
week = product_sales[['week', 'years_as_customer', 'revenue', 'state']]
week.head()

In [ ]:
hello = week.groupby(['week', 'years_as_customer', 'state'])['revenue'].agg(['sum', 'median', 'mean'])

hello.head()